WordPress/gutenberg
# Use npm workspaces for packages

See this Slack comment:  https://wordpress.slack.com/archives/C02QB2JS7/p1727315839205439

> Does anyone know the best way to update existing dependencies?
> I'm referring to [this documentation](https://github.com/WordPress/gutenberg/blob/trunk/packages/README.md#updating-existing-dependencies), but the `lerna add {library} packages/{package-name}` command doesn't seem to work now.
> 
> ```
> lerna add change-case packages/a11y
> ERR! add The "add" command was removed by default in v7, and is no longer maintained.
> ERR! add Learn more about this change at https://lerna.js.org/docs/legacy-package-management
> ```
> 
> The npm command doesn't work either.
> 
> ```
> npm install change-case -w packages/components
> npm error No workspaces found:
> npm error   --workspace=packages/components
> ---
> npm install change-case -w components
> npm error No workspaces found:
> npm error   --workspace=components
> ```

## What?

This PR adds npm workspaces and updates the documentation to allow existing dependencies to be properly updated.

## Why?

lerna has a `lerna add` command to add dependencies to the monorepo. However, since npm native now has a similar feature, [this command was removed in version 7](https://lerna.js.org/docs/legacy-package-management). Therefore, the current documented method cannot properly update dependencies.

## How?

For npm publishing, we will continue to use lerna as before. This PR allows us to run npm native commands that are an alternative to `lerna add` by adding a `workspaces` to `package.json`.

lerna will refer to `workspaces` in `package.json` by default, but since the settings in `lerna.json` take precedence, this should not affect npm publishing ([Source](https://lerna.js.org/docs/api-reference/configuration#packages)).

## Testing Instructions

Please run a command like the following to make sure the dependencies are added properly:

```
npm install change-case -w packages/a11y
```

The script modules webpack build config should be identical. This can be verified by adding some logging and executing the file `node ./tools/webpack/script-modules.js`:

```diff
diff --git before/tools/webpack/script-modules.js after/tools/webpack/script-modules.js
index 3733b8954f..4d57d1e0dd 100644
--- before/tools/webpack/script-modules.js
+++ after/tools/webpack/script-modules.js
@@ -101,3 +101,5 @@ module.exports = {
 		aggregateTimeout: 500,
 	},
 };
+
+console.log( module.exports );
```

Closes #59933.
